protocols/autonatv2/tests/autonatv2.rs

```diff
@@ -0,0 +1,516 @@
+                    tested_addr: Some(tested_addr),
+                    data_amount,
+                    server,
+                    result: Err(_),
+                })) if tested_addr == bob_unreachable_address && server == alice_peer_id => {
+                    Some(data_amount)
+                }
+                _ => None,
+            })
+            .await;
+        bob_done_tx.send(()).unwrap();
+        data_amount
+    };
+
+    let (alice_data_amount, bob_data_amount) = tokio::join!(alice_task, bob_task);
+    assert_eq!(alice_data_amount, bob_data_amount);
+    handler.abort();
+}
+
+async fn new_server() -> Swarm<CombinedServer> {
+    let mut node = Swarm::new_ephemeral(|identity| CombinedServer {
+        autonat: libp2p_autonatv2::server::Behaviour::default(),
+        identify: libp2p_identify::Behaviour::new(libp2p_identify::Config::new(
+            "/libp2p-test/1.0.0".into(),
+            identity.public().clone(),
+        )),
+    });
+    node.listen().with_tcp_addr_external().await;
+
+    node
+}
+
+async fn new_client() -> Swarm<CombinedClient> {
+    let mut node = Swarm::new_ephemeral(|identity| CombinedClient {
+        autonat: libp2p_autonatv2::client::Behaviour::new(
+            OsRng,
+            Config::default().with_recheck_interval(Duration::from_nanos(0)),
+        ),
+        identify: libp2p_identify::Behaviour::new(libp2p_identify::Config::new(
+            "/libp2p-test/1.0.0".into(),
+            identity.public().clone(),
+        )),
+    });
+    node.listen().with_tcp_addr_external().await;
+    node
+}
+
+#[derive(libp2p_swarm::NetworkBehaviour)]
+#[behaviour(prelude = "libp2p_swarm::derive_prelude")]
+struct CombinedServer {
+    autonat: libp2p_autonatv2::server::Behaviour,
+    identify: libp2p_identify::Behaviour,
+}
+
+#[derive(libp2p_swarm::NetworkBehaviour)]
+#[behaviour(prelude = "libp2p_swarm::derive_prelude")]
+struct CombinedClient {
+    autonat: libp2p_autonatv2::client::Behaviour,
+    identify: libp2p_identify::Behaviour,
+}
+
+async fn new_dummy() -> Swarm<libp2p_identify::Behaviour> {
+    let mut node = Swarm::new_ephemeral(|identity| {
+        libp2p_identify::Behaviour::new(libp2p_identify::Config::new(
+            "/libp2p-test/1.0.0".into(),
+            identity.public().clone(),
+        ))
+    });
+    node.listen().with_tcp_addr_external().await;
+    node
+}
+
+async fn bootstrap() -> (Swarm<CombinedServer>, Swarm<CombinedClient>) {
+    let mut alice = new_server().await;
+    let cor_server_peer = alice.local_peer_id().clone();
+    let mut bob = new_client().await;
+    let cor_client_peer = bob.local_peer_id().clone();
+
+    let bob_external_addrs = Arc::new(bob.external_addresses().cloned().collect::<Vec<_>>());
+    let alice_bob_external_addrs = bob_external_addrs.clone();
+
+    bob.connect(&mut alice).await;
+
+    let (tx, rx) = oneshot::channel();
+
+    let alice_task = async move {
```

Comment:
Why use `move`? That just makes things unnecessarily hard because you lose ownership of `alice`. If you just create a regular future via `futures::future::join()`, you can do all of this using borrowing and don't need the channels :)

Reply:
Yes you are right, I will refactor.